Feat(eos_designs): EVPN Multicast L3 (OISM) Support (#2276)
aristanetworks/ansible-avd: ansible_collections/arista/avd/roles/eos_designs/doc/network-services.md

>       underlay_l2_multicast_group_ipv4_pool: < IPv4_address/Mask >

Reviewer: There's no default specified? (line 141)

Author: Not my code, not this PR, but see Claus' comments elsewhere!

Reviewer 2: It is ok not to have a default value.

      underlay_l2_multicast_group_ipv4_pool_offset: < int >

    # Enable L3 Multicast for all SVIs and l3vlans within Tenant | Optional
    # - In the evpn-l3ls design type, this enables L3 EVPN Multicast (aka OISM)
    # - Multicast group binding for VRF is created only for Multicast traffic. BULL traffic will use ingress-replication

>     # - Configures binding between VXLAN, VLAN, and multicast group ipv4 address using the following formula:

Reviewer: Shouldn't all official standards/protocols etc. be named using upper case instead of all lower case? E.g. IPv4 vs ipv4 and IGMP vs igmp and so on?

Author: Agreed; have fixed this in the L2, L3 multicast and IGMP querier sections.  This probably needs doing throughout the documents at some point.

    #   < l3_multicast.evpn_underlay_l3_multicast_group_ipv4_pool > + < vrf_vni - 1 > + < l3_multicast.evpn_underlay_l3_multicast_group_ipv4_pool_offset >.
    # - The recommendation is to assign a /20 block within the 232.0.0.0/8 Source-Specific Multicast range.
    # - If enabled on an SVI using the anycast default gateway feature, a diagnostic loopback (see below) MUST be configured to source IGMP traffic.
    # - Enables `evpn multicast` on the router bgp VRF.
    # - When enabled on an SVI:
    #     - If switch is part of an MLAG pair, enables "pim ipv4 sparse-mode" on the SVI.
    #     - If switch is standalone or A-A MH, enables "ip igmp" on the SVI.
    #     - If "ip address virtual" is configured, enables "pim ipv4 local-interface" and uses the diagnostic Loopback defined in the VRF
    l3_multicast:

>       enabled: < true | false | default false >

Reviewer: isn't is written everywhere else as "< true | false | default -> false >"?

Reviewer 2: no need to show default when we default to false

Author: Fixed, but again was just copying behaviour seen elsewhere.

      evpn_underlay_l3_multicast_group_ipv4_pool: < IPv4_address/Mask > # Required
      evpn_underlay_l3_multicast_group_ipv4_pool_offset: < int > # Optional
      evpn_peg:
        # For each group of nodes, allow configuration of EVPN PEG options | Optional

>         # At least one RP address must be configured for EVPN PEG to be configured.

Reviewer 2: Since these options are not required for PEG to be configured at all, maybe we should change the wording and move it somewhere else. PEG will be configured if at least one RP address is configured _and_ PIM is enabled on l3_interfaces (or?)

Author: Maybe evpn_peg_options?  There are other options we could configure here (around PEG DR selection algorithms for example), but just haven't done so yet (would require eos_cli_config_gen dev work too).

Reviewer 2: I think the key `evpn_peg` is fine. My concern was more on the remarks here. Maybe move line 164 down to the l3_interfaces, where you state that you configure PEG.

Author: Have done so.

        - nodes: [ < node_1 >, < node_2 >, < node_N > ]                # Optional - will apply to all nodes with RP addresses configured if not set.
          transit: < true | false | default false >                    # Enable EVPN PEG transit mode
      rp_addresses:
        # For each group of nodes, allow configuration of RP Addresses & associated groups

>         - nodes: [ < node_1 >, < node_2 >, < node_N > ]                # Optional - will apply to all nodes if not set.

Reviewer 2: let's swap so `rps` is the first option, since it is required and `nodes` is optional.

Author: Done

      underlay_l2_multicast_group_ipv4_pool: < IPv4_address/Mask >
      underlay_l2_multicast_group_ipv4_pool_offset: < int >

    # Enable L3 Multicast for all SVIs and l3vlans within Tenant | Optional
    # - In the evpn-l3ls design type, this enables L3 EVPN Multicast (aka OISM)
    # - Multicast group binding for VRF is created only for Multicast traffic. BULL traffic will use ingress-replication
    # - Configures binding between VXLAN, VLAN, and multicast group ipv4 address using the following formula:
    #   < l3_multicast.evpn_underlay_l3_multicast_group_ipv4_pool > + < vrf_vni - 1 > + < l3_multicast.evpn_underlay_l3_multicast_group_ipv4_pool_offset >.
    # - The recommendation is to assign a /20 block within the 232.0.0.0/8 Source-Specific Multicast range.

>     # - If enabled on an SVI using the anycast default gateway feature, a diagnostic loopback (see below) must be configured to source IGMP traffic.

Reviewer 3: ```suggestion
    # - If enabled on an SVI using the anycast default gateway feature, a diagnostic loopback (see below) MUST be configured to source IGMP traffic.
```

>     # - Enables "evpn multicast" on the router bgp VRF.

Reviewer 3: ```suggestion
    # - Enables `evpn multicast` on the router bgp VRF.
```

    # - When enabled on an SVI:
    #     - If switch is part of an MLAG pair, enables "pim ipv4 sparse-mode" on the SVI.
    #     - If switch is standalone or A-A MH, enables "ip igmp" on the SVI.
    #     - If "ip address virtual" is configured, enables "pim ipv4 local-interface" and uses the diagnostic Loopback defined in the VRF
    l3_multicast:
      enabled: < true | false | default false >
      evpn_underlay_l3_multicast_group_ipv4_pool: < IPv4_address/Mask > # Required
      evpn_underlay_l3_multicast_group_ipv4_pool_offset: < int > # Optional
      node_settings: # Optional
        # For each group of nodes, allow configuration of EVPN PEG features | Optional
        # and allow configuration of specific RPs and associated multicast groups
        # Designed to be used in conjunction with enabling PIM on l3_interfaces to make the device a PEG (see below)

>         # An individual node should be in one node group only

Reviewer 3: is the code checking for this? (I guess the `break` will prevent the code to be executed for both)

Author: Correct; the break means that only one set of settings will apply.

        # This will create a loopback with virtual source-nat enable to perform diagnostics from the switch.
        vtep_diagnostic:

>           # Loopback interface number | Required (when vtep_diagnostics defined)

Reviewer 3: ```suggestion
          # Loopback interface number | Required (when vtep_diagnostic defined)
```

            - < hostname1 >
            - < hostname2 >

>         # Explicitly enable or disable l3_multicast to override setting of tenants.<tenant>.l3_multicast.enabled.

Reviewer 4: ```suggestion
        # Explicitly enable or disable evpn_l3_multicast to override setting of tenants.<tenant>.evpn_l3_multicast.enabled.
```

>         # Allow override of tenants.<tenant>.l3_multicast.node_settings

Reviewer 4: ```suggestion
        # Allow override of tenants.<tenant>.evpn_l3_multicast.node_settings
```

        evpn_l3_multicast:
          enabled: < true | false >
          evpn_peg:
            # For each group of nodes, allow configuration of EVPN PEG features | Optional
            - nodes: [ < node_1 >, < node_2 >, < node_N > ]                # Optional - will apply to all nodes with RP addresses configured if not set.
              transit: < true | false | default false >                    # Enable EVPN PEG transit mode

>         rp_addresses:                                                    # For each group of nodes, allow configuration of RP Addresses & associated groups

Reviewer 2: Should be updated similar to tenant config. `pim_rp_addresses`

Author: Thanks, my documentation update was really sloppy - my bad!

            - < hostname1 >
            - < hostname2 >

        # Explicitly enable or disable l3_multicast to override setting of tenants.<tenant>.l3_multicast.enabled.
        # Allow override of tenants.<tenant>.l3_multicast.node_settings
        l3_multicast:
          enabled: < true | false >
          evpn_peg:
            # For each group of nodes, allow configuration of EVPN PEG features | Optional

>             # At least one RP address must be configured for EVPN PEG to be configured.

Reviewer 2: remove this line, since it will be moved to l3_interfaces.

Author: Done

            - < hostname1 >
            - < hostname2 >

        # Explicitly enable or disable l3_multicast to override setting of tenants.<tenant>.l3_multicast.enabled.
        # Allow override of tenants.<tenant>.l3_multicast.node_settings

>         l3_multicast:
>           enabled: < true | false >
>           node_settings: # Optional
>             # For each group of nodes, allow configuration of EVPN PEG features | Optional
>             - nodes: [ < node_1 >, < node_2 >, < node_N > ]    # Optional
>               evpn_peg:                                        # Optional
>                 transit: < true | false | default false >      # Enable EVPN PEG transit mode
>               rp_addresses:                                    # RP Addresses & associated groups
>                 - rps:                                         # A minimum of one RP must be specified if node_settings is defined
>                     - < rp_address_1 >
>                     - < rp_address_2 >
>                   groups:                                      # Optional
>                     - < group_prefix_1/mask >
>                     - < group_prefix_2/mask >
> 

Reviewer 2: Having a deep model of nested lists will make it harder to override only some settings. In your code you are testing for `node_settings` specifically, so everything below will override on VRF level.
I would prefer to split it up a bit, to also open for future support of adjacent features.
```suggestion
        l3_multicast:
          enabled: < true | false >
          evpn_peg:
            # For each group of nodes, allow configuration of EVPN PEG features | Optional
            # At least one RP address must be configured for EVPN PEG to be configured.
            - nodes: [ < node_1 >, < node_2 >, < node_N > ]                # Optional - will apply to all nodes with RP addresses configured if not set.
              transit: < true | false | default false >                    # Enable EVPN PEG transit mode

          rp_addresses:                                                    # For each group of nodes, allow configuration of RP Addresses & associated groups
            - nodes: [ < node_1 >, < node_2 >, < node_N > ]                # Optional - will apply to all nodes if not set.
              rps: [ < rp_address_1 >, < rp_address_2 > ]                  # A minimum of one RP must be specified
              groups: [ < group_prefix_1/mask >, < group_prefix_1/mask > ] # Optional

```

Author: Have now implemented this data model.

            evpn_l2_multicast:
              enabled: < true | false >

>             # Explicitly enable or disable l3_multicast to override setting of tenants.<tenant>.l3_multicast.enabled and

Reviewer 4: ```suggestion
            # Explicitly enable or disable evpn_l3_multicast to override setting of tenants <tenant>.evpn_l3_multicast.enabled and
```

>             # tenants.<tenant>.vrfs.<vrf>.l3_multicast.enabled

Reviewer 4: ```suggestion
            # tenants.<tenant>.vrfs.<vrf>.evpn_l3_multicast.enabled
```

>             l3_multicast:

Reviewer 4: ```suggestion
            evpn_l3_multicast:
```

                  hash_algorithm: < md5 | sha1 | sha256 | sha384 | sha512, Default -> sha512 >
                  key: < key password >

>             # Enable PIM sparse-mode on the interface; requires "l3_multicast" to be enabled on the VRF/Tenant

Reviewer 4: ```suggestion
            # Enable PIM sparse-mode on the interface; requires "evpn_l3_multicast" to be enabled on the VRF/Tenant
```

                  hash_algorithm: < md5 | sha1 | sha256 | sha384 | sha512, Default -> sha512 >
                  key: < key password >

            # Enable PIM sparse-mode on the interface; requires "l3_multicast" to be enabled on the VRF/Tenant

>             # Enabling this implicitly makes the device a PIM External Gateway (PEG)

Reviewer 2: This is only true for EVPN.

Author: Noted.
